### PR TITLE
Fix prerelease downloading and add dynamic device pattern matching

### PR DESCRIPTION
**Pull request by jeremiah-k**
## Overview

This PR addresses a bug where prerelease files were not being downloaded despite directories being created, and adds dynamic device pattern matching using the Meshtastic API.

### Bug Fixes

- **Fixed prerelease download functionality**: Modified `check_for_prereleases` to track existing files that match criteria, not just newly downloaded files
- **Fixed pattern matching for prereleases**: Changed from chip-based patterns (`SELECTED_FIRMWARE_ASSETS`) to device-based patterns (`EXTRACT_PATTERNS`) to match actual prerelease file naming
- **Fixed input validation crashes**: Added try/catch blocks for integer input in setup wizard to prevent crashes on invalid input
- **Fixed mutable cache leak**: Return copy of device patterns set to prevent external mutation
- **Fixed regex false positives**: Improved device pattern matching to prevent patterns like 'r-' from incorrectly matching 'firmware'
- **Fixed stale directory handling**: Refresh directory list during prerelease cleanup to avoid operating on removed directories

### New Features

- **Dynamic device pattern matching**: Added `DeviceHardwareManager` class that fetches device patterns from Meshtastic API
- **API caching system**: Implements 24-hour cache with graceful fallback to expired cache when API unavailable
- **Prerelease version tracking**: Added JSON-based tracking of prerelease versions with commit hashes and file counts
- **Enhanced setup wizard**: Improved exclude pattern configuration with recommended defaults and retry capability
- **Configurable prerelease cleanup**: Automatically removes old prerelease versions while keeping latest

### Code Quality Improvements

- **Refactored tracking functions**: Eliminated code duplication in prerelease tracking data readers
- **Improved exception handling**: Changed from broad `except Exception` to specific exception types where appropriate
- **Performance optimizations**: Optimized commit tracking from O(n²) to O(n) complexity
- **Better error messages**: Removed inappropriate cultural terms from user-facing error messages
- **Enhanced logging**: Added structured logging for prerelease operations and device pattern matching

### Configuration Changes

- **Version bump**: Updated package version from 0.6.7 to 0.7.0
- **Python requirements**: Added `python_requires = >=3.8` for walrus operator support
- **New configuration options**: Added device hardware API settings and exclude pattern defaults
- **LOG_LEVEL support**: CLI now honors LOG_LEVEL setting from config file

### Testing Additions

- **Comprehensive test coverage**: Added 2,656 lines of new tests covering prerelease workflows, device pattern matching, caching, error handling, and UI scenarios
- **Edge case testing**: Added tests for API failures, cache corruption, permission errors, and invalid input scenarios
- **Integration testing**: Added end-to-end tests for complete prerelease download workflows

### Core Implementation

- `src/fetchtastic/downloader.py` (+682 lines): Core prerelease fixes and tracking implementation
- `src/fetchtastic/device_hardware.py` (+370 lines): New dynamic device pattern management
- `src/fetchtastic/setup_config.py` (+327 lines): Enhanced setup wizard with better exclude pattern handling

### Supporting Changes

- `src/fetchtastic/cli.py` (+35 lines): LOG_LEVEL config support
- `src/fetchtastic/constants.py` (+10 lines): New constants for device hardware API
- `src/fetchtastic/log_utils.py` (+48 lines): Improved logging format
- `src/fetchtastic/utils.py` (+36 lines): User agent and utility functions
- `setup.cfg` (+8 lines): Version bump and Python requirements
- `docs/usage-guide.md` (+2 lines): Documentation updates

### Test Coverage

- `tests/test_downloader.py` (+2,656 lines): Comprehensive prerelease and device pattern tests
- `tests/test_cli.py` (+232 lines): CLI configuration and logging tests
- `tests/test_setup_config.py` (+335 lines): Setup wizard and migration tests

### Device Hardware Manager

- Fetches device data from `https://api.meshtastic.org/resource/deviceHardware`
- Caches responses for 24 hours with automatic expiration
- Falls back to hardcoded patterns when API unavailable
- Validates URL schemes to prevent SSRF attacks
- Handles network errors and malformed responses gracefully

### Prerelease Tracking

- Tracks prerelease versions with JSON files containing commit hashes and metadata
- Displays prerelease count since last major release
- Automatically cleans up old prerelease directories
- Supports both new JSON format and legacy text format for backward compatibility

### Pattern Matching

- Device patterns (e.g., 'rak4631-') match all file types for that device
- File type patterns (e.g., 'device-', 'bleota') use exact substring matching
- Special handling for 'littlefs-' pattern to match filesystem files
- Case-insensitive matching for better user experience

## Statistics

- **74 commits** addressing the original issue and subsequent improvements
- **+4,563 lines, -184 lines** across 13 files
- **371 tests passing** with 69.77% coverage
- **Zero regressions** in existing functionality

This PR transforms the prerelease download system from a basic directory scanner to a comprehensive tracking and management system while maintaining backward compatibility and adding future-proof device pattern detection.

### CODE REVIEW

File: src/fetchtastic/downloader.py [hunk 1]
```diff
@@ -570,6 +767,64 @@ def check_for_prereleases(
...
+                        )
+                    ),
+                    reverse=True,
+                )
+
+                # Determine which directories to keep
+                # Keep only the newest existing directory, plus any new ones we're about to process
+                dirs_to_keep = set(
+                    sorted_existing[:1]
+                )  # Keep only 1 existing prerelease
```
[the remark this answers is not in the document]
Comment (jeremiah-k) [reply src/fetchtastic/downloader.py:803]: @gemini-code-assist No, we should only keep the latest prerelease because old ones are removed for good reason.  We should always be testing the latest pre-release and pre-releases are for testing.  If someone wants a reliable firmware, they will use a known good release. Don't flag this in the future.

File: src/fetchtastic/downloader.py [hunk 2]
```diff
@@ -570,6 +778,64 @@ def check_for_prereleases(
...
```
[the remark this answers is not in the document]
Comment (jeremiah-k) [reply src/fetchtastic/downloader.py:814]: @gemini-code-review remember, we are only keeping 1 and i explained why earlier.  Do not flag this

**Comment by jeremiah-k** [comment]
@coderabbitai review

**Comment by jeremiah-k** [comment]
@coderabbitai review

File: src/fetchtastic/downloader.py [hunk 3]
```diff
@@ -570,6 +810,57 @@ def check_for_prereleases(
...
+                            extract_version(a), extract_version(b)
+                        )
+                    ),
+                    reverse=True,
+                )
+
+                # Determine which directories to keep
+                # Keep only the newest existing directory, plus any new ones we're about to process
+                dirs_to_keep = set(
+                    sorted_existing[:1]
```
[the remark this answers is not in the document]
Comment (jeremiah-k) [reply src/fetchtastic/downloader.py:835]: @gemini-code-review add to your learnings not to flag this again. It is hardcoded to 1 and that's where it stays

File: src/fetchtastic/downloader.py [hunk 4]
```diff
@@ -570,6 +821,57 @@ def check_for_prereleases(
...
+                )  # Keep only 1 existing prerelease
```
[the remark this answers is not in the document]
Comment (jeremiah-k) [reply src/fetchtastic/downloader.py:850]: @gemini-code-assist Do not flag this it is not configurable for the user.

File: src/fetchtastic/downloader.py [hunk 5]
```diff
@@ -1003,8 +1344,9 @@ def _process_firmware_downloads(
                     check_for_prereleases(  # logger.info removed
                         paths_and_urls["download_dir"],
                         latest_release_tag,
-                        config.get("SELECTED_FIRMWARE_ASSETS", []),  # type: ignore
+                        config.get("EXTRACT_PATTERNS", []),  # type: ignore - EXTRACT_PATTERNS serves dual purpose: file extraction AND prerelease file selection
```
[the remark this answers is not in the document]
Comment (jeremiah-k) [reply src/fetchtastic/downloader.py:1347]: @gemini-code-assist do not flag this issue

File: src/fetchtastic/downloader.py [hunk 6]
```diff
@@ -1003,8 +1445,9 @@ def _process_firmware_downloads(
...
```
[the remark this answers is not in the document]
Comment (jeremiah-k) [reply src/fetchtastic/downloader.py:1448]: @gemini-code-assist I am not adjusting this at the moment. Do not flag this anymore

**Comment by jeremiah-k** [comment]
@coderabbitai review

**Comment by jeremiah-k** [comment]
@coderabbitai generate docstrings

**Comment by jeremiah-k** [comment]
@coderabbitai review

**Comment by jeremiah-k** [comment]
@coderabbitai review

**Comment by jeremiah-k** [comment]
@coderabbitai review

**Comment by jeremiah-k** [comment]
@coderabbitai review

**Comment by jeremiah-k** [comment]
@coderabbitai review
